### PR TITLE
[lld/COFF] Handle -start-lib / -end-lib better in /reproduce: output

### PR DESCRIPTION
**Pull request by nico**
Previously, we'd collect all input files in Driver::filePaths, and then write filePaths after all other flags in
createResponseFile(). This meant that `-start-lib foo.obj -end-lib` would be written as `-start-lib -end-lib foo.obj`, changing semantics.

Instead, remove Driver::filePaths, and handle things that fed into it directly:

* OPT_INPUT is now handled in the same way as other flags, so that we now get `-start-lib foo.obj -end-lib` in response.txt as desired. Add a test for -start-lib / -end-lib and /reproduce:.

* OPT_wholearchive_file needs explicit handling now -- but before, this was buggy as well: We'd put the flag without a rewritten path in response.txt, but also the rewritten input file without wholearchive semantics via filePaths. So this commit makes --whole-archive work with /reproduce: too, and adds test coverage.

* /defaultlib:foo is now written as /defaultlib:foo into response.txt, instead of writing the resolved path previously. While response.txt looks slightly differently, both should have the same semantics, and this should be mostly a no-op. (It does require updating a test.)

* /defaultlib: from .drectve sections are no longer recorded in response.txt. This seems like a progression -- in the non-repro case they come from .obj files, so they should come (only) from there in the repro case too. This adds test coverage for this case.

Makes createResponseFile() look more like the versions in the ELF and MachO ports too.

### CODE REVIEW

File: lld/test/COFF/linkrepro.test [hunk 1]
```diff
@@ -96,3 +106,90 @@ Test /call-graph-ordering-file (can't be used with /order:, needs separate test)
...
+
+#--- archive.s
+	.text
+	.intel_syntax noprefix
+	.globl	exportfn3
+	.p2align	4
+exportfn3:
+	ret
+
+	.section	.drectve,"yni"
```
Comment (mstorsjo) [inline lld/test/COFF/linkrepro.test:194]: What's the section type `yni` here, vs the type `yn` in `drectve.s` above?

**Review by mstorsjo — APPROVED** [review]
LGTM, this looks like a nice simplification! I wonder why it wasn't done like this from the get-go.

I remember running into some problems with some arguments around reproduce outputs as well, let's see if I remember them, now when you're digging around here :-)

With PDB outputs, I think the repro output has the problem that it will try to write the output into a directory that might not exist; this probably will need to be rewritten into something like just `/pdb:out.pdb` rather than the original PDB output path.

Then for really tricky things; if you have static libraries involved, that actually are thin libraries, i.e. just referencing other object files on disk, those aren't included in the repro package. And even if they would be, the static library would need to be rewritten to reference them with their relocatable name within the repro bundle.

File: lld/test/COFF/linkrepro.test (hunk 1 above)
Comment (nico) [reply lld/test/COFF/linkrepro.test:194]: I got `"yni"` from `clang-cl /FA` output, while `"yn"` is from another test. Both should work :)

**Comment by nico** [comment]
> LGTM, this looks like a nice simplification! 

Thanks!
 
> With PDB outputs, I think the repro output has the problem that it will try to write the output into a directory that might not exist; this probably will need to be rewritten into something like just `/pdb:out.pdb` rather than the original PDB output path.

That sounds right.

> Then for really tricky things; if you have static libraries involved, that actually are thin libraries, i.e. just referencing other object files on disk, those aren't included in the repro package. And even if they would be, the static library would need to be rewritten to reference them with their relocatable name within the repro bundle.

lld/test/ELF/reproduce-thin-archive.s lld/test/MachO/reproduce-thin-archive-objc.s  lld/test/MachO/reproduce-thin-archives.s all exist, so someone seems to have thought about this in the past. We should probably do the same in the COFF port, if we aren't yet :)

**Comment by nico** [comment]
> lld/test/ELF/reproduce-thin-archive.s lld/test/MachO/reproduce-thin-archive-objc.s lld/test/MachO/reproduce-thin-archives.s all exist, so someone seems to have thought about this in the past. We should probably do the same in the COFF port, if we aren't yet :)

It me! :open_mouth: https://reviews.llvm.org/D92456

I think /reproduce sets up a sysroot, so maybe changing path isn't even necessary as long as we include the files? I don't remember what I was thinking back then, but I think it did work.

The behavior for thin archives without an index is also different across ports. If I find time, I'd like to look at that, so I might look at thin archives + lld/COFF too.

**Comment by nico** [comment]
> https://reviews.llvm.org/D92456

Future self (or whoever gets to it): Also do ##97169 for lld/COFF.

**Pull request merged**